src/main/java/com/hubspot/smtp/utils/ByteBufs.java

```diff
@@ -4,15 +4,112 @@
 import java.util.List;
 
 import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.CompositeByteBuf;
 import io.netty.buffer.Unpooled;
+import io.netty.util.ByteProcessor;
 
 public final class ByteBufs {
   private static final byte CR = '\r';
   private static final byte LF = '\n';
   private static final byte DOT = '.';
   private static final byte[] DOT_DOT = {DOT, DOT};
+  private static final byte[] NOT_CR_LF = {'x', 'x'};
   private static final byte[] CR_LF = {CR, LF};
   private static final byte[] EMPTY_BYTE_ARRAY = new byte[0];
+  private static final ByteBuf DOT_DOT_BUFFER = Unpooled.unreleasableBuffer(Unpooled.wrappedBuffer(DOT_DOT));
+  private static final ByteBuf CR_LF_BUFFER = Unpooled.unreleasableBuffer(Unpooled.wrappedBuffer(CR_LF));
+
+  /**
+   * Returns a {@link CompositeByteBuf} that contains the same data as {@code sourceBuffer}, but with
+   * SMTP dot-stuffing applied, and (if {@code} appendCRLF is true) a CRLF appended.
+   *
+   * <p>If dot-stuffing is not required, and {@code appendCRLF} is false, {@code sourceBuffer} is
+   * returned. In all other cases, {@code allocator} will be used to create a new {@code ByteBuf}
+   * with a {@code refCnt} of one.
+   *
+   * <p>The {@code previousBytes} parameter is used to maintain dot-stuffing across a series
+   * of buffers. Pass the last two bytes of a previous buffer here to ensure an initial dot
+   * will be escaped if necessary. Passing null indicates this is the first or only buffer
+   * for this message.
+   *
+   * @param allocator the {@code ByteBufAllocator} to use for new {@code ByteBuf}s
+   * @param sourceBuffer the source message data
+   * @param previousBytes the previous two bytes of the message, or null
+   * @param appendCRLF whether to append CRLF to the end of the returned buffer
+   */
+  public static ByteBuf createDotStuffedBuffer(ByteBufAllocator allocator, ByteBuf sourceBuffer, byte[] previousBytes, boolean appendCRLF) {
```

Comment:
Yeah, I think we can probably remove the overload of `createDotStuffedBuffer` that takes a `byte[]` and move this somewhere in the messages package.